### ISSUE TITLE
SqlException: 'String or binary data would be truncated in table' while SaveToken() for OAuth2

### ISSUE THREAD

**Issue by kirati-exxonmobil** [umbraco/Umbraco.AuthorizedServices#40]
### System
**Umbraco** version: 10.2.1
**Umbraco.AuthorizedServices** version: 0.3.2

---
### Issue
I added a new authorized service in `appsettings.json` named `Survey` aim to call external API with OAuth2 authorization code.
When I run the project and navigate to backoffice `Settings` > `Authorized Services` > `Survey` then click `Authorize Service` button, I got the exception.
```
SqlException: String or binary data would be truncated in table 'umbraco-cms.dbo.umbracoAuthorizedServiceOAuth2Token', column 'accessToken'. Truncated value: 'CfDJ8ACj27kvevhCo6u3x6aXer_xxxxxxxxx'. The statement has been terminated.
```
I looked to the call stack seem like the app try to save a token in the database table `umbracoAuthorizedServiceOAuth2Token` with column `accessToken` after received the token.

I'm not sure this is a bug or I just did something wrong.

Any idea ?

---
### Additional info
**appsettings.json**, [remove my actual value]
```
    "AuthorizedServices": {
      "TokenEncryptionKey": "",
      "Services": {
        "survey": {
          "DisplayName": "Survey",
          "AuthenticationMethod": "OAuth2AuthorizationCode",
          "IdentityHost": "https://login.microsoftonline.com",
          "TokenHost": "https://login.microsoftonline.com",
          "RequestIdentityPath": "/[my tenant]/oauth2/v2.0/authorize",
          "RequestTokenPath": "/[my tenant]/oauth2/v2.0/token",
          "RequestTokenFormat": "FormUrlEncoded",
          "JsonSerializer": "SystemTextJson",
          "AuthorizationUrlRequiresRedirectUrl": true,
          "SampleRequest": "/identity/me"
          "ApiHost": "https://[my api url]/api",
          "ClientId": "[my client]",
          "ClientSecret": "[my secret]",
          "Scopes": "api://[my scope uri]/user_impersonation"
        },
```

**Call stack**
```
Microsoft.Data.SqlClient.SqlException (0x80131904): String or binary data would be truncated in table 'umbraco-cms.dbo.umbracoAuthorizedServiceOAuth2Token', column 'accessToken'. Truncated value: 'CfDJ8ACj27kvevhCo6u3x6aXer_xxxxxxxxx'.
The statement has been terminated.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean isAsync, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String method)
   at Microsoft.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String methodName)
   at Microsoft.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Umbraco.Cms.Infrastructure.Persistence.FaultHandling.RetryPolicy.ExecuteAction[TResult](Func`1 func)
   at NPoco.Database.ExecuteNonQueryHelper(DbCommand cmd)
   at NPoco.Database.InsertAsyncImp[T](PocoData pocoData, String tableName, String primaryKeyName, Boolean autoIncrement, T poco, Boolean sync)
   at NPoco.Database.Insert[T](String tableName, String primaryKeyName, Boolean autoIncrement, T poco)
   at NPoco.Database.Insert[T](T poco)
   at Umbraco.AuthorizedServices.Services.Implement.DatabaseOAuth2TokenStorage.SaveToken(String serviceAlias, OAuth2Token token)
   at Umbraco.AuthorizedServices.Services.Implement.AuthorizedServiceBase.StoreOAuth2Token(String serviceAlias, OAuth2Token token)
   at Umbraco.AuthorizedServices.Services.Implement.AuthorizedServiceAuthorizer.SendRequest(ServiceDetail serviceDetail, Dictionary`2 parameters, Boolean isExchangeTokenRequest)
   at Umbraco.AuthorizedServices.Services.Implement.AuthorizedServiceAuthorizer.AuthorizeOAuth2AuthorizationCodeServiceAsync(String serviceAlias, String authorizationCode, String redirectUri, String codeVerifier)
   at Umbraco.AuthorizedServices.Controllers.AuthorizedServiceResponseController.HandleOAuth2IdentityResponse(String code, String state)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Umbraco.Cms.Web.Common.Middleware.BasicAuthenticationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass6_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Umbraco.Cms.Web.BackOffice.Middleware.BackOfficeExternalLoginProviderErrorMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass6_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext httpContext, Boolean retry)
   at StackExchange.Profiling.MiniProfilerMiddleware.Invoke(HttpContext context) in C:\projects\dotnet\src\MiniProfiler.AspNetCore\MiniProfilerMiddleware.cs:line 119
   at Umbraco.Cms.Web.Common.Middleware.UmbracoRequestMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Umbraco.Cms.Web.Common.Middleware.UmbracoRequestMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass6_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Umbraco.Cms.Web.Common.Middleware.PreviewAuthenticationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass6_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Umbraco.Cms.Web.Common.Middleware.UmbracoRequestLoggingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next)
   at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass6_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
ClientConnectionId:394be641-72e1-4fab-83d7-ff6ffe5ed3db
Error Number:2628,State:1,Class:16
```

**Comment by AndyButland** [comment]
It sounds like the access token the service you are using is generating is too long for what we have allocated in table storage.  Access tokens have been assumed to be less than 1000 characters.  Are you able to confirm that the token you receive is longer than that?  And if so, can you let us know how long so we can appropriately increase the database column size?  Thanks.

**Comment by kirati-exxonmobil** [comment]
> It sounds like the access token the service you are using is generating is too long for what we have allocated in table storage. Access tokens have been assumed to be less than 1000 characters. Are you able to confirm that the token you receive is longer than that? And if so, can you let us know how long so we can appropriately increase the database column size? Thanks.

@AndyButland You are right. In my case, the Access token contains 1867 characters in total.

**Comment by AndyButland** [comment]
OK, I've created a PR to extend the column size.  Once that's reviewed we'll create a new release of the package to address this issue.

Meantime if you want to move forward now there's no harm in you already increasing the field size in your own database to 2000 characters via:

```
ALTER TABLE umbracoAuthorizedServiceOAuth2Token
ALTER COLUMN accessToken nvarchar(2000) NOT NULL

ALTER TABLE umbracoAuthorizedServiceOAuth2Token
ALTER COLUMN refreshToken nvarchar(2000) NOT NULL
GO
```

**Comment by kirati-exxonmobil** [comment]
@AndyButland Thank you for your suggestion, I tried to increase accessToken size to 2000 characters as your suggest but still failed. I tried again to increase it to 3000 characters and now it is successfully save the token and call to  sample request via back office :). I checked the token length in the database, seem like the actual token length is 2608 charactors.

Is it possible to extend the column size to 3000 characters ?
